### ISSUE TITLE
add instructions on developing with Docker

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#8340]
At https://mail.python.org/pipermail/scipy-dev/2018-January/022372.html there's a discussion on how to make it easier to get started with SciPy development. One thing that came out is that it would be valuable to have clear instructions on how to set up a development environment with Docker.

The best place is probably to add it to https://docs.scipy.org/doc/scipy/reference/dev/index.html and link it from http://scipy.github.io/devdocs/hacking.html#recommended-development-setup

**Comment by andyfaff** [comment]
I wonder if there was a specific Docker container for scipy development whether it'd be worth doing the travis CI with that container as well.

A.

**Comment by rgommers** [comment]
I think it could be one config in the build matrix, but not the only one. With the CI setup we should aim to cover as wide a range of normal install usage patterns as possible, and the average user won't install with Docker.

**Comment by andyfaff** [comment]
@mdhaber, where you developing documentation about this? If not, then I can write information up on this, but my RST is pretty bad, much better at MD.

**Comment by mdhaber** [comment]
@andyfaff I intend to, but I am taking a break from it, so it would be awesome if you could start. If you do, please consider how it will fit with the existing development documentation. It is great if you want to work with a non-Anaconda Python, but it would be nice if you could also include an Anaconda option.

If you want to draft in MD on the Wiki, I can help you convert to RST. I did a ton of that for the documentation overhaul. RST is a huge pain and I wouldn't want you to have to suffer through it. Please avoid nested markup; RST can't handle it. It is OK to include all links with URLs, but if the link is SciPy-internal it would be nice if you can link to the source document so it is easier for me to convert it to an internal reference.

**Comment by andyfaff** [comment]
@mdhaber, I've written a [draft](https://github.com/scipy/scipy/wiki/Development-Environment-Quickstart-Guide-(Docker)-(WIP)). It should be workable at the moment, but would benefit from someone that's not familiar with Docker to go through it.
I understand the rationale for giving Anaconda instructions, but using a Docker image which already has Python3.6 and 3.7 reduces the need for that.

**Comment by andyfaff** [comment]
Another reason it's not really worth including virtualenv/conda instructions with a Docker setup is because starting/stopping a container is cheap, and the only persistent changes are those made to a mounted volume.

**Comment by mdhaber** [comment]
Downloading Docker now. Can you install SuiteSparse?

**Comment by andyfaff** [comment]
Sure, I'll do that later today.

**Comment by mdhaber** [comment]
I ran into:
https://github.com/docker/for-win/issues/3174
Can't create a local user on this computer, so I'll have to wait until I get home.

I'll edit this post later, but some thoughts for my reference:
* more direct link to install instructions? (maybe not)
* I ran into https://github.com/docker/for-win/issues/1825
* how to tell that Docker is running?
* how long should step 4 take? did it hang?

**Comment by mdhaber** [comment]
I ran into this error in step 8:
```
root@a94blahblah:/home/scipy# python3.7 setup.py build_ext --inplace
Traceback (most recent call last):
  File "setup.py", line 34, in <module>
    from numpy.distutils.core import setup
  File "/usr/local/lib/python3.7/dist-packages/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/usr/local/lib/python3.7/dist-packages/numpy/core/__init__.py", line 100, in <module>
    from . import _internal
  File "/usr/local/lib/python3.7/dist-packages/numpy/core/_internal.py", line 11, in <module>
    import platform
  File "/usr/lib/python3.7/platform.py", line 116, in <module>
    import sys, os, re, subprocess
  File "/usr/lib/python3.7/subprocess.py", line 50, in <module>
    import signal
  File "/home/scipy/signal/__init__.py", line 289, in <module>
    from . import sigtools, windows
  File "/home/scipy/signal/windows/__init__.py", line 41, in <module>
    from .windows import *
  File "/home/scipy/signal/windows/windows.py", line 9, in <module>
    from scipy import linalg, special, fft as sp_fft
```

**Comment by andyfaff** [comment]
I am using Docker for macOS, so I don't know the ins and outs of Docker for Windows. It would certainly be a pain point using these instructions if it's difficult to get Docker running on your computer. I could imaging installing Docker without admin rights is likely to be painful.

If you are getting issues mounting a drive in the container, then you can remove the `-v $PWD/scipy:/home/scipy` part of the docker run command. If you do that then the scipy directory won't be available in `/home/scipy`. The container does have git available in it, so you could `git clone` when inside the container.

> more direct link to install instructions? (maybe not)

 I think that detailed instructions for getting Docker installed and running are outside the scope of the tutorial, because they will never be truly all-encompassing, and will quickly become out of date. That's why I linked to the Docker documentation. If there is a better link for install instructions, then lets change it.

> how to tell that Docker is running? system tray.

Not sure on windows. But from a CMD/powershell window you should be able to use `docker info` and that would give information. On macOS there is an icon in the system tray.

> I ran into docker/for-win#1825

THis is an issue with getting Docker setup correctly. I don't think we should put instructions for this kind of stuff in our documentation. However, if there is a FAQ/frequently encountered issues then we could link to that.

**Comment by andyfaff** [comment]
> I ran into this error in step 8:

I updated the Docker image at `scipy/scipy-dev`:
```
docker run -it --rm scipy/scipy-dev
```
It should work if you install:
```
pip3.7 install numpy cython pytest pybind11
```

**Comment by mdhaber** [comment]
Great. Seems to be building.

**Comment by mdhaber** [comment]
All tests passing! Very nice. 
What commands have to be run next time I want to start up the container? I imagine the dependencies get reinstalled, but SciPy wouldn't need to be re-built if the source hasn't changed?

**Comment by andyfaff** [comment]
> What commands have to be run next time I want to start up the container?

You use the same command to start the container.

>  I imagine the dependencies get reinstalled,

Everything you've done in the container gets cleared when you exit the container. Thus all the packages you install via pip (pytest, numpy, pybind11, cython) have to be reinstalled. Everytime you start the container it'll be in exactly the same state. If one wants to add extra packages into the container to be present when the container starts, then the image has to be built from an amended Dockerfile. I didn't include the dependencies in the Dockerfile because they will change version number from time to time.

>  but SciPy wouldn't need to be re-built if the source hasn't changed?

In this example the only state that persists after you shut the container down are the changes that were made to the mounted scipy directory. The -v flag makes the a directory from the host OS available in the container. When you build scipy Python amends the files in the scipy directory, making `build, `dist` directories, etc. Any changes that are made in that mounted directory are still present when the container is shut down, and would be available the next time you started the container (assuming that you mounted the directory again) Thus, a re-build would take a lot less time the next time you issued a build command inside the container.

> Also, how would you typically "install" it so scipy can be imported from a different working directory?

This question isn't totally clear to me. When the container starts up you can install scipy by using the pip command, with pip grabbing scipy from PyPi. Alternatively you could do a `pip3.7 install .` within the scipy directory for scipy to be installed from source into site-packages. After changing out of the scipy directory you can then start an interpreter up and start working with it.

**Comment by mdhaber** [comment]
That's what I meant, right. By 'install', I meant any procedure that allows you to `import scipy` from another working directory. The downside of the first solution is that it is not the development version of SciPy. The issue with the second is that you have to install every time you make changes. I was wondering whether you add the development version of SciPy to the PYTHONPATH manually, or if there is something similar to `conda develop` which adds it when you start a given virtual environment. `python setup.py develop` might work, too, but I've not tried it. If you don't do anything like that in your setup, no problem.

**Comment by andyfaff** [comment]
I'm only familiar with `pip install .` from the scipy directory, which makes it useable from a different working directory.
I don't know enough about how other modes work (`python setup.py develop`, `pip install -e .`) to comment more deeply.

**Comment by mdhaber** [comment]
OK, when I convert to RST I might add a note about:
```
export PYTHONPATH="/home/scipy:$PYTHONPATH"
source ~/.bashrc
```
After running that, you can import scipy from any working directory.

Then again, whether that is useful depends on how the user intends to interact with the container. If all goes well, I won't use the container for anything except building SciPy and running a Spyder kernel, then I can connect to that kernel from Spyder running outside the container. I am hoping that will do all my development as though the kernel were running locally (outside the container). Does that make sense?

Unfortunately, I'm unable to launch a Spyder kernel inside the container. I did:
```
pip install paramiko spyder-kernels spyder
```
but then:
```
python3.7 -m spyder_kernels.console`
```
gives me
```
/usr/bin/python3.7: Error while finding module specification for 'spyder_kernels.console' (ModuleNotFoundError: No module named 'spyder_kernels')
```
Do you use Spyder? Or do you use some other IDE that would allow connecting to a kernel running in the container?

**Comment by mdhaber** [comment]
BTW I rebased on master and now when I try to build inside the container, I get a familiar error:
```
compiling Fortran sources
Fortran f77 compiler: /usr/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -fPIC -O3 -funroll-loops
Fortran f90 compiler: /usr/bin/gfortran -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
Fortran fix compiler: /usr/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
compile options: '-DNO_ATLAS_INFO=-1 -Ibuild/src.linux-x86_64-3.7/scipy/linalg -I/usr/local/lib/python3.7/dist-packages/numpy/core/include -I/usr/include/python3.7m -c'
gfortran:f77: /home/scipy/scipy/_build_utils/src/wrap_dummy_g77_abi.f
gfortran:f77: scipy/linalg/_fblas-f2pywrappers.f
/usr/bin/gfortran -Wall -g -Wall -g -shared build/temp.linux-x86_64-3.7/scipy/linalg/_fblasmodule.o build/temp.linux-x86_64-3.7/build/src.linux-x86_64-3.7/scipy/linalg/fortranobject.o build/temp.linux-x86_64-3.7/home/scipy/scipy/_build_utils/src/wrap_dummy_g77_abi.o build/temp.linux-x86_64-3.7/scipy/linalg/_fblas-f2pywrappers.o -L/usr/lib/x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/7 -L/usr/lib/gcc/x86_64-linux-gnu/7 -Lbuild/temp.linux-x86_64-3.7 -llapack -lf77blas -lcblas -latlas -lf77blas -lcblas -lgfortran -o scipy/linalg/_fblas.cpython-37m-x86_64-linux-gnu.so -Wl,--version-script=build/temp.linux-x86_64-3.7/link-version-scipy.linalg._fblas.map
/usr/bin/ld: cannot open output file scipy/linalg/_fblas.cpython-37m-x86_64-linux-gnu.so: No such file or directory
collect2: error: ld returned 1 exit status
error: Command "/usr/bin/gfortran -Wall -g -Wall -g -shared build/temp.linux-x86_64-3.7/scipy/linalg/_fblasmodule.o build/temp.linux-x86_64-3.7/build/src.linux-x86_64-3.7/scipy/linalg/fortranobject.o build/temp.linux-x86_64-3.7/home/scipy/scipy/_build_utils/src/wrap_dummy_g77_abi.o build/temp.linux-x86_64-3.7/scipy/linalg/_fblas-f2pywrappers.o -L/usr/lib/x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/7 -L/usr/lib/gcc/x86_64-linux-gnu/7 -Lbuild/temp.linux-x86_64-3.7 -llapack -lf77blas -lcblas -latlas -lf77blas -lcblas -lgfortran -o scipy/linalg/_fblas.cpython-37m-x86_64-linux-gnu.so -Wl,--version-script=build/temp.linux-x86_64-3.7/link-version-scipy.linalg._fblas.map" failed with exit status 1
```

**Comment by mdhaber** [comment]
That was in a shell that didn't have administrator privileges. When I retry with admin rights, I now get:
```
root@582c44750712:/home/scipy# python3.7 setup.py build_ext --inplace
Traceback (most recent call last):
  File "setup.py", line 34, in <module>
    from numpy.distutils.core import setup
  File "/usr/local/lib/python3.7/dist-packages/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/usr/local/lib/python3.7/dist-packages/numpy/core/__init__.py", line 100, in <module>
    from . import _internal
  File "/usr/local/lib/python3.7/dist-packages/numpy/core/_internal.py", line 11, in <module>
    import platform
  File "/usr/lib/python3.7/platform.py", line 116, in <module>
    import sys, os, re, subprocess
  File "/usr/lib/python3.7/subprocess.py", line 50, in <module>
    import signal
  File "/home/scipy/signal/__init__.py", line 289, in <module>
    from . import sigtools, windows
  File "/home/scipy/signal/windows/__init__.py", line 41, in <module>
    from .windows import *
  File "/home/scipy/signal/windows/windows.py", line 9, in <module>
    from scipy import linalg, special, fft as sp_fft
ModuleNotFoundError: No module named 'scipy'
```

**Comment by andyfaff** [comment]
> OK, when I convert to RST I might add a note about:

Those tweaks could be added to the Dockerfile so that they'd be present in the container when it starts up. However, if the user doesn't mount a scipy directory, then that would fail. As you say, everyone will have a slightly different way of working with the code, so my intention was to provide a minimal base for people to build/compile the code.

> If all goes well, I won't use the container for anything except building SciPy and running a Spyder kernel, then I can connect to that kernel from Spyder running outside the container. I am hoping that will do all my development as though the kernel were running locally (outside the container). Does that make sense?

It does make sense. There are lots of docker images designed for interacting with [Jupyter notebooks](https://github.com/jupyter/docker-stacks). It looks as if [people want to do the same thing](https://github.com/spyder-ide/spyder/issues/7147) with Spyder. Docker images for that kind of thing have to expose ports from the container and connect them to the host, as well as start various services (e.g. ssh, webservers, a Jupyter notebook server, etc). Nothing like that is setup in the Docker image I created. It's main purpose is to setup the build dependencies required for scipy. It would be possible to take a minimal Jupyter notebook [instance](https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile) and build [FROM](https://docs.docker.com/engine/reference/builder/#from) that, just adding other build dependencies like suitesparse, build-essentials, etc. As soon as you go down that route the image size can start to inflate and things do get more complicated.

> Do you use Spyder? Or do you use some other IDE that would allow connecting to a kernel running in the container? It would be really nice if users can write and run code from a GUI just as though SciPy were installed locally (outside the container).

I don't use Spyder myself, I use the Pycharm IDE for things like that. I do all my developing in macOS. My development workflow consists of making changes to source code in Pycharm, possibly experimenting in Jupyter to figure things out, then running the test suite to check that everything is fine. Once it's working on my computer I'll make a PR, or run the test suite in Docker to see if the changes work on other OS'. If they work in macOS, but not other OS' then I'll fire up a virtual machine and do some debugging in those to see if things work. That doesn't usually happen for scipy, but is a common occurrence for a PyQt application that I work on.

**Comment by andyfaff** [comment]
You edited your first post, was that edit made before or after your second post (which failed on the  ModuleNotFoundError)?
I've not experienced the issues you're coming across here, so I don't know how to advise. Given that the container is always exactly the same state when it starts up, I'd say that the issue might be in one of two places:
1) When you mount the scipy directory in the container, the container needs to have write access to that directory to be able to do an --inplace build. This might require some kind of admin privilege when starting the container.
2) Given that you share the scipy directory between the host and container its state may become inconsistent. If you do a `git clean -xdf` (read up on that command before issuing it, clears the repository of things that aren't tracked by the repo) then you might find problems go away.

**Comment by mdhaber** [comment]
> As soon as you go down that route the image size can start to inflate and things do get more complicated.

That's something I meant to ask about. When I first read about Docker, it sounded like a Dockerfile was a plain text file, and I could build an image from that. In that case, the download would be quite small. But I noticed that the file on scipy's Dockerhub is quite large, so I assume that is some sort of image. Can we distribute a Dockerfile instead, if image size is a concern? Perhaps there is a tradeoff between image size and setup time - can we provide a basic image, and also provide a Dockerfile for the rest?  (Thanks for trying to understand. I realize that this might not make sense if I'm not understanding how Docker works.)

> My development workflow consists of making changes to source code in Pycharm, possibly experimenting in Jupyter to figure things out,

Making changes and testing in GUIs is really helpful for development work. I think that the ability to do that is very important, and would be worth a large image size.

>You edited your first post, was that edit made before or after your second post (which failed on the ModuleNotFoundError)?

Edit was after the second post. 
For the first post, I was using PowerShell without admin rights.
For the second post, I was using PowerShell with admin rights.
For the update to the first post, I was using PowerShell without admin rights.
I just tried again with admin rights. Same error. Weird.

>If you do a git clean -xdf

Yes, I have been making sure that the working directory is clean. I've run into issues with this sort of thing in the past, but thanks for the reminder.

**Comment by mdhaber** [comment]
TLDR:
In this adventure of building SciPy in a Linux environment from my Windows computer, I often get different behaviors running tests from command line vs running them from within Python. Any thoughts on why?

I thought I'd mention that I tried installing miniconda and building SciPy with conda Python and conda-installed build-time dependencies. It all went OK, and tests run almost without issue from the command line. (Just one failure I've seen before - [gh-10550](https://github.com/scipy/scipy/issues/10550).

But when I run the tests from within Python, (`import scipy; scipy.test()`), I get a segfault! 

```
scipy/linalg/tests/test_basic.py .......................................................s....................... [ 16%]
.......................................................                                                          [ 16%]
scipy/linalg/tests/test_blas.py .......Fatal Python error: Segmentation fault

Current thread 0x00007f3a63c84740 (most recent call first):
  File "/home/scipy/scipy/linalg/tests/test_blas.py", line 158 in test_complex_dotu
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/python.py", line 165 in pytest_pyfunc_call
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/python.py", line 1451 in runtest
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/runner.py", line 117 in pytest_runtest_call
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/runner.py", line 192 in <lambda>
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/runner.py", line 220 in from_call
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/runner.py", line 192 in call_runtest_hook
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/runner.py", line 167 in call_and_report
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/runner.py", line 87 in runtestprotocol
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/runner.py", line 72 in pytest_runtest_protocol
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/main.py", line 278 in pytest_runtestloop
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/main.py", line 257 in _main
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/main.py", line 213 in wrap_session
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/main.py", line 250 in pytest_cmdline_main
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/root/miniconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/root/miniconda3/lib/python3.7/site-packages/_pytest/config/__init__.py", line 74 in main
  File "/home/scipy/scipy/_lib/_testutils.py", line 69 in __call__
  File "<stdin>", line 1 in <module>
Segmentation fault
```

**Comment by andyfaff** [comment]
> When I first read about Docker, it sounded like a Dockerfile was a plain text file

It is, see the Dockerfile [PR](https://github.com/scipy/scipy/pull/10655/files) I suggest scipy starts off with.

> and I could build an image from that. 

You can build an image from the file by running `docker build .` if the Dockerfile is in the current working directory. The docker build process then has to download the operating system image, then make the changes requested in the docker file, such as installing compilers, etc. The image that's made is going to (typically) be at least 100 Mb, probably much more. You can get lightweight images based on Alpine Linux that are only 5Mb. But as soon as you install compilers, etc, everything balloons.

>  But I noticed that the file on scipy's Dockerhub is quite large,

I propose to add the Docker file to the scipy repo (small). However, images (large) are stored on Dockerhub. There are pro's and con's to either: downloading a Dockerfile and building it yourself, or docker grabbing an image and running it.

**Comment by mdhaber** [comment]
Thanks. 

I'll stop adding stuff to this issue and convert those instructions to RST. I would like to add instructions for conveniently interacting with this version of SciPy, but we can continue that discussion in a separate issue.

**Comment by andyfaff** [comment]
Building the image can require lots of bandwidth, and it takes time. Downloading the image when it's run also takes bandwidth. However, images are cached on computer, so there is no download required the next time. The bandwidth required to download an image, as opposed to the bandwidth required to build an image won't be orders of magnitude apart.

**Comment by mdhaber** [comment]
Note to self: Docker Desktop doesn't run on Windows home. Instead, use the legacy [Docker Toolbox](https://docs.docker.com/toolbox/toolbox_install_windows/).
